File: repositories/refresh_token_repository.go
package repositories

import (
	"context"
	"errors"

	"doctor-service/entities"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type RefreshTokenRepository interface {
	Create(ctx context.Context, token *entities.RefreshToken) error
	FindByHash(ctx context.Context, tokenHash string) (*entities.RefreshToken, error)
	Revoke(ctx context.Context, tokenHash string) error
	RevokeAllForDoctor(ctx context.Context, doctorID uuid.UUID) error
}

type GormRefreshTokenRepository struct {
	db *gorm.DB
}

func NewRefreshTokenRepository(db *gorm.DB) RefreshTokenRepository {
	return &GormRefreshTokenRepository{db: db}
}

func (r *GormRefreshTokenRepository) Create(ctx context.Context, token *entities.RefreshToken) error {
	return r.db.WithContext(ctx).Create(token).Error
}

func (r *GormRefreshTokenRepository) FindByHash(ctx context.Context, tokenHash string) (*entities.RefreshToken, error) {
	var token entities.RefreshToken
	err := r.db.WithContext(ctx).Where("token_hash = ?", tokenHash).First(&token).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &token, err
}

func (r *GormRefreshTokenRepository) Revoke(ctx context.Context, tokenHash string) error {
	return r.db.WithContext(ctx).Model(&entities.RefreshToken{}).Where("token_hash = ?", tokenHash).Update("is_revoked", true).Error
}

func (r *GormRefreshTokenRepository) RevokeAllForDoctor(ctx context.Context, doctorID uuid.UUID) error {
	return r.db.WithContext(ctx).Model(&entities.RefreshToken{}).Where("doctor_id = ?", doctorID).Update("is_revoked", true).Error
}

File: python_backend/repositories/refresh_token_repository.py
from typing import Optional
from uuid import UUID
from sqlalchemy.orm import Session
from entities.refresh_token import RefreshToken

class RefreshTokenRepository:
    def __init__(self, db: Session):
        self.db = db

    def Create(self, token: RefreshToken) -> None:
        self.db.add(token)
        self.db.commit()
        self.db.refresh(token)

    def FindByHash(self, token_hash: str) -> Optional[RefreshToken]:
        return self.db.query(RefreshToken).filter(RefreshToken.token_hash == token_hash).first()

    def Revoke(self, token_hash: str) -> None:
        self.db.query(RefreshToken).filter(RefreshToken.token_hash == token_hash).update({"is_revoked": True})
        self.db.commit()

    def RevokeAllForDoctor(self, doctor_id: UUID) -> None:
        self.db.query(RefreshToken).filter(RefreshToken.doctor_id == doctor_id).update({"is_revoked": True})
        self.db.commit()
